tests.cpp

```diff
@@ -5,456 +5,800 @@
+template<typename T>
+std::ostream& print_cont(std::ostream& os, T const& v){
+    os << "[ ";
+
+    for (int x = 0; x < v.size(); x++)
+    {
+        os << v[x];
+        if (x != v.size() - 1)
+        {
+            os << ", ";
+        }
+    }
+
+    os << " ]";
+
+    return os;
+}
+
+template<typename T, size_t N>
+std::ostream& operator<<(std::ostream& os, static_vector<T, N> const& v){
+    return print_cont(os, v);
+}
+
+template<typename T, size_t N>
+std::ostream& operator<<(std::ostream& os, std::array<T, N> const& v){
+    return print_cont(os, v);
+}
+
+std::ostream& operator<<(std::ostream& os, char c){
+    return os.operator<<(+c);
+}
+
+void debug_print_impl()
+{
+}
+
+template<typename First, typename... T>
+void debug_print_impl(First const& f, T const&... t)
+{
+    std::cerr << f;
+    if (sizeof...(T) != 0)
+    {
+        std::cerr << ", ";
+    }
+    debug_print_impl(t...);
+}
+
+template<typename... T>
+void debug_print(T const&... t)
+{
+    if (sizeof...(T) != 0)
+    {
+        std::cerr << "(";
+    }
+    debug_print_impl(t...);
+    if (sizeof...(T) != 0)
+    {
+        std::cerr << ")";
+    }
+}
+
+template<typename T>
+auto debug_transform(T const& t)
+{
+    return t;
+}
+
+template<typename... T>
+void assert_failure(const char* expression, const char* file, long line, T const&... t) {
+    std::cerr << "Assertion failure: " << expression;
+    debug_print(t...);
+    std::cerr << " failed at " << file
+              << ':' << line << ".\n";
+    std::exit(1); // Exit program early
+}
 
 #define ASSERT(e) ((e) ? true : (assert_failure(#e, __FILE__, __LINE__), false))
+#define ASSERT_MESSAGE(e, ...) ((e) ? true \
+        : (assert_failure(#e " note (" #__VA_ARGS__ ") == ", __FILE__, __LINE__, __VA_ARGS__), false))
+
+#define ASSERT_EQUAL(a, b) ASSERT_MESSAGE(a == b, a, b)
+#define ASSERT_UNEQUAL(a, b) ASSERT_MESSAGE(a != b, a, b)
+
+
+template<typename... T>
+std::array<char, sizeof...(T)> make_c_array(T&&... t)
+{
+    return { static_cast<char>(t)... };
+}
 
 // Self-referential object that tests whether copies are semantically correct,
 // using the copy constructors of stored objects.
-struct Copyable {
-    Copyable() : self(this) { constructed_++; }
-    Copyable(const Copyable& other) : self(other.verify() ? this : nullptr) {
-        constructed_++;
+//
+// It also stores a payload in order to allow us to treat it "like" an int or
+// whatever in order to allow us to verify orders
+template<typename T = char>
+struct Copyable_t {
+    Copyable_t() : self(this) {
+        ASSERT(constructed_.find(this) == std::end(constructed_));
+        constructed_.insert(this);
     }
-    Copyable& operator=(const Copyable& other) {
+
+    template<typename U,
+        typename=std::enable_if_t<
+            std::is_same<std::decay_t<U>, T>::value
+        >
+    >
+    Copyable_t(U&& t) : Copyable_t() {
+        inner = std::forward<U>(t);
+    }
```

Comment:
```suggestion
    Copyable_t(T t) : Copyable_t() {
        inner = std::move(t);
    }
```

I think this can be simplified a bit for the test code. Or do we need this constructor to be a template?